stackslib/src/net/unsolicited.rs

```diff
@@ -0,0 +1,1030 @@
+            Ok(Some(block_height)) => block_height,
+            Ok(None) => {
+                debug!(
+                    "{:?}: We already know the inventory state in {} for {}",
+                    &self.local_peer, outbound_neighbor_key, consensus_hash
+                );
+                return Ok(None);
+            }
+            Err(NetError::NotFoundError) => {
+                // is this remote node simply ahead of us?
+                if let Some(convo) = self.peers.get(&event_id) {
+                    if self.chain_view.burn_block_height < convo.burnchain_tip_height {
+                        debug!("{:?}: Unrecognized consensus hash {}; it is possible that {} is ahead of us", &self.local_peer, consensus_hash, outbound_neighbor_key);
+                        return Err(NetError::NotFoundError);
+                    }
+                }
+                // not ahead of us -- it's a bad consensus hash
+                debug!("{:?}: Unrecognized consensus hash {}; assuming that {} has a different chain view", &self.local_peer, consensus_hash, outbound_neighbor_key);
+                return Ok(None);
+            }
+            Err(NetError::InvalidMessage) => {
+                // punish this peer
+                info!(
+                    "Peer {:?} sent an invalid update for {}",
+                    &outbound_neighbor_key,
+                    if microblocks {
+                        "streamed microblocks"
+                    } else {
+                        "blocks"
+                    }
+                );
+                self.bans.insert(event_id);
+
+                if let Some(outbound_event_id) = self.events.get(&outbound_neighbor_key) {
+                    self.bans.insert(*outbound_event_id);
+                }
+                return Ok(None);
+            }
+            Err(e) => {
+                warn!(
+                    "Failed to update inv state for {:?}: {:?}",
+                    &outbound_neighbor_key, &e
+                );
+                return Ok(None);
+            }
+        };
+        Ok(Some(block_sortition_height))
+    }
+
+    /// Buffer a message for re-processing once the burnchain view updates.
+    /// If there is no space for the message, then silently drop it.
+    fn buffer_data_message(&mut self, event_id: usize, msg: StacksMessage) {
+        let Some(msgs) = self.pending_messages.get_mut(&event_id) else {
+            self.pending_messages.insert(event_id, vec![msg]);
+            debug!(
+                "{:?}: Event {} has 1 messages buffered",
+                &self.local_peer, event_id
+            );
+            return;
+        };
+
+        // check limits against connection opts, and if the limit is not met, then buffer up the
+        // message.
+        let mut blocks_available = 0;
+        let mut microblocks_available = 0;
+        let mut blocks_data = 0;
+        let mut microblocks_data = 0;
+        let mut nakamoto_blocks_data = 0;
```

Comment:
The way this is currently structured, if _any_ limit is met, it stops buffering any messages of any type. Does this work because `msgs` is always filled with one type? Or is this the intended behavior?

Reply:
Yes, I believe that's an untested regression.  Will fix.